### ISSUE TITLE
Train NVS using Sockeye 

### ISSUE THREAD

**Issue by HossamAmer12** [awslabs/sockeye#1063]
Hi, I am new to sockeye and I would like to only get the learnable parameters of NVS as done in your paper:
[[The Devil is in the Details: On the Pitfalls of Vocabulary Selection in Neural Machine Translation](https://aclanthology.org/2022.naacl-main.136.pdf)](https://aclanthology.org/2022.naacl-main.136/
)

I have an enfr trained transformer checkpoint (*.pt mode), which I'd like to freeze its weights. I have wmt17 enfr data with its vocab in sentence piece.

After evaluating this checkpoint, I generated my train data stored in 2 pickle files (train.pickle and train_labels.pickle). The train includes the encoder outs and train_labels includes the label data.

With this information, is there a way to only learn the parameters of NVS (W and b in your paper) using sockeye? 

Any guidance would be appreciated.

**Comment by tdomhan** [comment]
To train NVS you can use this branch:
https://github.com/awslabs/sockeye/tree/naacl2022

NVS is enabled via the flag `--neural-vocab-selection` to Sockeye. If you have a trained model then you can use the `--params` flag to load the existing (Sockeye) parameters. I assume the model that you already trained is a Sockeye model, right?

Training data can either be specified as text files via the `-s` and `-t` flags to `sockeye.train` or you could run `sockeye.prepare_data` to convert training data into binary format. There's no need to dump encoder states. You train the NMT model with NVS enabled on text files that contain source/target text.

I hope this clarifies things. If not let me know.

**Comment by HossamAmer12** [comment]
Thanks, @tdomhan.. Your response is very helpful .... Can you explain what is _num_target_vocabs_ in https://github.com/awslabs/sockeye/blob/94cdad722cd9d0b2d24ddd401ec170db1280b83b/test/unit/test_loss.py

**Comment by HossamAmer12** [comment]
I figured out the answers to the first question @tdomhan... In the sockeye repo, you have multiple target vocabs.. so normally I should set it to 1.

The second one could be inspired by the masking in the transformer -- but I am not sure as there's no mention in the paper. Can you please explain?

The third one is still not clear to me.

**Comment by tdomhan** [comment]
Multiple target vocabularies only come into play when you use target factors. The first vocabulary is for the main output symbol while the others could be POS tags or other target factors. Normally this would be 1 indeed.

Can you explain what you mean by the target attention masks? The link code contains the target bag of words (all target words will be 1 everything else 0. In general we will only have a loss for tokens other than the PAD token, maybe this is what you are referring to).


> Also, I see that there is C.NVS_TYPE_LOGIT_MAX and C.NVS_TYPE_EOS? 
To reproduce the paper you should use the default, C.NVS_TYPE_LOGIT_MAX. The function takes the encoder outputs as input. 

Hope that helps :)

**Comment by HossamAmer12** [comment]
1) Got it... Thank you @tdomhan 
2) In the following code:
https://github.com/awslabs/sockeye/blob/94cdad722cd9d0b2d24ddd401ec170db1280b83b/sockeye/nvs.py#L35

There is `att_mask`... What does this variable indicate? Does this just exclude pad tokens?

3) Great... I assume that the GT input to this loss function would be  [target tokens + EOS token]... Correct?

**Comment by tdomhan** [comment]
Oh ok. So that is to not make predictions for PAD tokens and only for real tokens.

So the input is always all the tokens including PAD, but the loss will not be computed for PAD positions.

**Comment by HossamAmer12** [comment]
Thanks @tdomhan for your help .. Just to make sure, if our GT target tokens s this:

 ```
tensor([[ 1628,   385,   159,  6737, 28890,    78,    87,   166, 31935,    42,
           868,  1329,    87,   166, 31935,   100,   107,  5790,    25, 31789,
         16769,     2,     1,     1,     1,     1,     1,     1,     1,     1,
             1,     1,     1,     1,     1,     1,     1,     1,     1,     1,
             1,     1,     1,     1,     1,     1,     1,     1,     1,     1,
             1,     1,     1,     1,     1,     1,     1,     1,     1,     1,
             1,     1,     1,     1,     1,     1,     1,     1,     1,     1,
             1,     1,     1,     1,     1,     1,     1,     1,     1,     1,
             1,     1,     1,     1,     1,     1,     1,     1,     1,     1,
             1,     1,     1,     1]], device='cuda:0') 
torch.Size([1, 94])
```

label_pred = [ 1628,   385,   159,  6737, 28890,    78,    87,   166, 31935,    42,
           868,  1329,    87,   166, 31935,   100,   107,  5790,    25, 31789,
         16769,     2,     1]
         
Yes? Later you will convert that to one hot representation

attn_mask will be what? -inf for all ones in the vector above?

**Comment by HossamAmer12** [comment]
Thanks @tdomhan  ... I think that I got good grasp of your answer.

So, in your linear encode vocab projection, you set the padding idx to -inf in order to exclude the PAD token from the predictions. When you compute the loss, this PAD token will already be set to 0. Therefore, the PAD token will also be excluded from the loss calculations. Please correct me if I am wrong.

Just one last question: Using your method, what was roughly the percentage of the vocab tokens that were active at inference time? What was your typical acceptance eval threshold?

**Comment by tdomhan** [comment]
> Just one last question: Using your method, what was roughly the percentage of the vocab tokens that were active at inference time? What was your typical acceptance eval threshold?

This would be what Fig. 3 and Fig. 4 show in the paper depending on the exact threshold used. The x-axis shows the number of items that are above the threshold, so the average vocabulary size.

Let me know if there is anything else that's unclear or that I missed above.

**Comment by HossamAmer12** [comment]
Thanks, @tdomhan ... Your clarifications are very helpful :)
When you do Lnvs + Lmt during training:
1) Do you apply the nvs threshold to Lmt and compute the loss based on the new logits? or do you use the logits from the model?
2) Do you divide Lnvs + Lmt by the train sample size (N) to compute the loss used for gradient update? (i.e (Lnvs + Lmt)/N )
3) I notice that Lnvs is normally in the range of 0.4-0.7 in the train start... is that what you observed in your experiments?
4) Why did you not use Lnvs only to train?

**Comment by HossamAmer12** [comment]
I got the answers to these questions... Will write them here in case anyone would like to read this chat :)

1. In training, the nvs threshold is not applied... Raw logits are taken. 
2. Loss is normalized by the input batch for gradient computation
3. That's usually the case... Lmt is dominating the total loss
4. That's open for experiments

Thanks, @tdomhan 👍
